packages/model-viewer/src/three-components/ARRenderer.ts

```diff
@@ -248,117 +327,330 @@ export class ARRenderer extends EventDispatcher {
+   * the input's y-coordinate). The XZ placement is found by first putting the
+   * scene's target at the hit point, drawing a ray from the camera to the
+   * target, and finding the XZ-intersection of this ray with the model's
+   * bounding box. The scene is then translated on the XZ plane to position this
+   * intersection point at the input hit point. If the ray does not intersect,
+   * the target is left at the hit point.
+   *
+   * This ensures the model is placed according to the chosen target, is not
+   * reoriented, and does not intersect the camera even when the model
+   * is large (unless the target is chosen outside of the model's bounding box).
    */
-  processXRInput(frame: XRFrame) {
-    const {session} = frame;
+  placeModel(hit: Vector3) {
+    const scene = this[$presentedScene]!;
+    const {model} = scene;
+    const {min, max} = model.boundingBox;
+
+    this[$placementBox]!.show = true;
+
+    const goal = this[$goalPosition];
+    goal.copy(hit);
+    const floor = hit.y;
+
+    const origin = this.camera.position.clone();
+    const direction = hit.clone().sub(origin).normalize();
+    // Pull camera back enough to be outside of large models.
+    origin.sub(direction.multiplyScalar(model.idealCameraDistance));
+    const ray = new Ray(origin, direction.normalize());
+
+    const model2World = this[$initialModel2World];
+    const modelPosition =
+        new Vector3().setFromMatrixPosition(model2World).add(hit);
+    model2World.setPosition(modelPosition);
+    const world2Model = new Matrix4().getInverse(model2World);
+    ray.applyMatrix4(world2Model);
+
+    // Make the box tall so that we don't intersect the top face.
+    max.y += 10;
+    ray.intersectBox(model.boundingBox, modelPosition);
+    max.y -= 10;
+
+    if (modelPosition != null) {
+      modelPosition.applyMatrix4(model2World);
+      goal.add(hit).sub(modelPosition);
+    }
+
+    // Move the scene's target to the model's floor height.
+    const target = scene.getTarget();
+    scene.setTarget(target.x, min.y, target.z);
+    // Ignore the y-coordinate and set on the floor instead.
+    goal.y = floor;
 
-    // Get current input sources. For now, only 'screen' input is supported,
-    // which is only added to the session's active input sources immediately
-    // before `selectstart` and removed immediately after `selectend` event.
-    // If we have a 'screen' source here, it means the output canvas was
-    // tapped.
-    const sources = Array.from(session.inputSources)
-                        .filter(input => input.targetRayMode === 'screen');
+    this.dispatchEvent({type: 'modelmove'});
+  }
 
-    if (sources.length === 0) {
+  [$onSelectStart](event: XRInputSourceEvent) {
+    const hitSource = this[$transientHitTestSource];
+    if (hitSource == null) {
       return;
     }
+    const fingers = this[$frame]!.getHitTestResultsForTransientInput(hitSource);
```

Comment:
Seems like this should be `null` checked

Reply:
nah, this is like a bunch of the `| null` members in here. It gets definitively set at the top of `$onWebXRFrame` which runs before any of this does. I'd love to get rid of the whole `| null` system that was in here, but I couldn't come up with anything better.